File: edition/2017/java/Data Structures/binary_tree.java
class Node {    

     Node left, right;
     int data;
 
     public Node() {
         left = null;
         right = null;
         data = 0;
     }

     public BTNode(int n) {
         left = null;
         right = null;
         data = n;
     }

     public void setLeft(Node n) {
         left = n;
     }

     public void setRight(Node n) {
         right = n;
     }

     public Node getLeft() {
         return left;
     }

     public Node getRight() {
         return right;
     }

     public void setData(int d) {
         data = d;
     }

     public int getData() {
         return data;
     }     
 }
 
 class BinaryTree {

     private Node root;
 
     public BinaryTree() {
         root = null;
     }

     public boolean isEmpty() {
         return root == null;
     }
    
     public void insert(int data) {
         root = insert(root, data);
     }
     
     private Node insert(Node node, int data)
     {
         if (node == null)
             node = new Node(data);
         else
         {
             if (node.getRight() == null)
                 node.right = insert(node.right, data);
             else
                 node.left = insert(node.left, data);             
         }
         return node;
     }     

     public int countNodes() {
         return countNodes(root);
     }
   
     private int countNodes(Node r) {
         if (r == null)
             return 0;
         else
         {
             int l = 1;
             l += countNodes(r.getLeft());
             l += countNodes(r.getRight());
             return l;
         }
     }     
 }

File: cpp/data_structures/binaryTree.cpp
#include <bits/stdc++.h>
using namespace std;

struct node{
	int data;
	node* right;
	node* left;
};

node* newnode(int data)
{
	node* node = new node;
	node -> data = data;
	node-> left = NULL;
	node -> right = NULL;
	return node;	
}
void height(node* root)
{
	if(root == NULL)
	   return 0;
	else
	{
	 int lheight = height(root->left);
	 int rheight = height(root-> right);
	 if(lheight > rheight)
	    return (lheight+1);
	 else 
	   return (rheight+1);
	}
}

int main()
{
	struct node* root = newnode(1);
	root-> left = newnode(2);
	root-> right = newnode(3);
	root -> left->left = newnode(4);
	getchar();
	return 0;
}

